include/openenclave/bits/sgx/sgxproperties.h

```diff
@@ -223,6 +231,54 @@ typedef struct _oe_sgx_enclave_properties
     FAMILY_ID,                                                            \
     ALLOW_DEBUG,                                                          \
     true,                                                                 \
+    0,                                                                    \
+    HEAP_PAGE_COUNT,                                                      \
+    STACK_PAGE_COUNT,                                                     \
+    TCS_COUNT)
+
+/**
+ * Defines the SGX2 properties for an enclave
+ *
+ * The enclave properties should only be defined once for all code compiled into
+ * an enclave binary. These properties can be overwritten at sign time by
+ * the oesign tool.
+ *
+ * @param[in] PRODUCT_ID ISV assigned Product ID (ISVPRODID) to use in the
+ * enclave signature
+ * @param[in] SECURITY_VERSION ISV assigned Security Version number (ISVSVN)
+ * to use in the enclave signature
+ * @param[in] EXTENDED_PRODUCT_ID ISV assigned Extended Product ID (ISVEXTPRODID)
+ *  to use in the enclave signature (SGX2 feature)
+ * @param[in] FAMILY_ID ISV assigned Product Family ID (ISVFAMILYID)
+ * to use in the enclave signature (SGX2 feature)
+ * @param[in] ALLOW_DEBUG If true, allows the enclave to be created with
+ * OE_ENCLAVE_FLAG_DEBUG and debugged at runtime
+ * @param[in] CAPTURE_PF_GP_EXCEPTIONS If true, allows the enclave to capture #PF/#GP
+ * exceptions if CPU supports the feature. The setting is ignored otherwise
+ * (SGX2 feature)
+ * @param[in] HEAP_PAGE_COUNT Number of heap pages to allocate in the enclave
+ * @param[in] STACK_PAGE_COUNT Number of stack pages per thread to reserve in
+ * the enclave
+ * @param[in] TCS_COUNT Number of concurrent threads in an enclave to support
+ */
+ #define OE_SET_ENCLAVE_SGX2(                                             \
```

Comment:
A developer might want use KSS and PF_GP_EXCEPTIONS. This macro does not support that.

Reply:
`EXTENDED_PRODUCT_ID` and `FAMILY_ID` are KSS-specific properties. The KSS flag should be set if any of these properties is set. This is the case for oesign: https://github.com/openenclave/openenclave/blob/d5155f48fc9958b017654998e1adf1fdf629906b/tools/oesign/oesign.c#L465.
If this is not the case for this macro, we should have a separate PR to fix it.

Reply:
I've updated the PR to expose the `REQUIRE_KSS` option in the new macro. However, I found that there is an inconsistency between using oesign + config file versus using MACRO, which should require a fix.

In the oesign + config scenario, the oesign will patch the oeinfo section based on the config file. For the case of KSS, the `SGX_FLAG_KSS` flag will be set if `EXTENDED_PRODUCT_ID` and `FAMILY_ID` is specified in the config file.

For the macro scenario, the `SGX_FLAG_KSS` is explicitly set based on the `REQUIRE_KSS` flag. If the `REQUIRE_KSS` flag is set to `true`, the `SGX_FLAG_KSS` will be set in the oeinfo section. Now even if the enclave is signed by oesign with both `EXTENDED_PRODUCT_ID` and `FAMILY_ID` unspecified in the config file, the `SGX_FLAG_KSS` will remain `true` in the oeinfo section.

Regarding the macro design, based on the current implementation, it may not be possible to support what you suggested. An `OE_SET_ENCLAVE_SGX*` macro, internally invoked the `_OE_SET_ENCLAVE_SGX_IMPL` macro, which effectively initializes the data structure of the oeinfo section at compile time. The struct initialization cannot be done separately, unless we have newly added features defined in separated structs and append them to the oeinfo structure. This requires some work to the loader stack and will not be needed by the oesign + config scenario. Do you think it's worth reworking the marco? Also note that the use of marco should be mainly for debug purposes (e.g., we only use it in our tests) and we should discourage using it in production enclaves.

Reply:
@mingweishih, thanks for the explanation of the challenges to implement a scalable MACRO solution. I think it's OK to start with the current implementation in the CapturePFGPExceptions PR and explore better solutions in future PRs.